lib/client/conntest/externalcloudaudit.go

```diff
@@ -0,0 +1,185 @@
+	cfg ExternalAuditStorageConnectionTesterConfig
+}
+
+// NewDatabaseConnectionTester returns a new DatabaseConnectionTester.
+func NewExternalAuditStorageConnectionTester(cfg ExternalAuditStorageConnectionTesterConfig) (*ExternalAuditStorageConnectionTester, error) {
+	return &ExternalAuditStorageConnectionTester{
+		cfg,
+	}, nil
+}
+
+// TestConnection tests the current configured ExternalCloudAudit draft by:
+// * Uploading a dummy file to both the audit events and session recordings S3 Buckets.
+// * Tests get object on the session recordings bucket.
+// * Tests the retrieval of the Glue table.
+// * Runs a test query against the audit events bucket through Athena.
+func (s *ExternalAuditStorageConnectionTester) TestConnection(ctx context.Context, req TestConnectionRequest) (types.ConnectionDiagnostic, error) {
+	if req.ResourceKind != types.KindExternalCloudAudit {
+		return nil, trace.BadParameter("invalid value for ResourceKind, expected %q got %q", types.KindExternalCloudAudit, req.ResourceKind)
+	}
+
+	connectionDiagnosticID := uuid.NewString()
+	connectionDiagnostic, err := types.NewConnectionDiagnosticV1(
+		connectionDiagnosticID,
+		map[string]string{},
+		types.ConnectionDiagnosticSpecV1{
+			// We start with a failed state so that we don't need to set it to each return statement once an error is returned.
+			// if the test reaches the end, we force the test to be a success by calling
+			// 	connectionDiagnostic.SetMessage(types.DiagnosticMessageSuccess)
+			//	connectionDiagnostic.SetSuccess(true)
+			Message: types.DiagnosticMessageFailed,
+		},
+	)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	if err := s.cfg.UserClient.CreateConnectionDiagnostic(ctx, connectionDiagnostic); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	// Test Connection to S3 Buckets
+	diag, diagErr, err := s.handleBucketsTest(ctx, connectionDiagnosticID)
+	if err != nil || diagErr != nil {
+		return diag, diagErr
+	}
+
+	// Test Connection to Glue Table
+	diag, diagErr, err = s.handleGlueTest(ctx, connectionDiagnosticID)
+	if err != nil || diagErr != nil {
+		return diag, diagErr
+	}
+
+	// Test Connection to Athena
+	diag, diagErr, err = s.handleAthenaTest(ctx, connectionDiagnosticID)
+	if err != nil || diagErr != nil {
+		return diag, diagErr
+	}
+
+	traceType := types.ConnectionDiagnosticTrace_CONNECTIVITY
+	const message = "External Audit Storage draft permissions are configured correctly."
+	connDiag, err := s.appendDiagnosticTrace(ctx, connectionDiagnosticID, traceType, message, nil)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	connDiag.SetMessage(types.DiagnosticMessageSuccess)
+	connDiag.SetSuccess(true)
+
+	if err := s.cfg.UserClient.UpdateConnectionDiagnostic(ctx, connDiag); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return connDiag, nil
+}
+
+func (s ExternalAuditStorageConnectionTester) handleBucketsTest(ctx context.Context, connectionDiagnosticID string) (types.ConnectionDiagnostic, error, error) {
```

Comment:
I'm thinking it through. We return early on either a diagnostic error or a test connection error. We wrap the test connection error or success into the diagnostic and return the diagnostic error, if any, that occurs when we perform the test connection.

If we aggregate the error then we could simplify the conditional check but it would cause the return error to be non-nil when the test connection was non-nil. We currently return just the diagErr because the test connection error is wrapped into the diagnostics. 

I don't know if `trace.NewAggregate` would allow us to return a nil error regardless of test connection, but then return the diagError if it occurs. 

If we are looking for simpler logic, I could create a struct like this:

```golang
type ConnectionError struct{
    DiagErr error
    Err error
}

func (c ConnectionError) Error() string {
    return trace.NewAggregate(c.DiagErr, c.Err)
}
```

and then use it like:

```golang
func (s ExternalAuditStorageConnectionTester) handleBucketsTest(ctx context.Context, connectionDiagnosticID string) (types.ConnectionDiagnostic, error) {
...
diag, err := s.handleBucketsTest(ctx, connectionDiagnosticID)
if err != nil {
    return diag, err.(ConnectionError).DiagErr
}

```

Reply:
Or just return the ConnectionError directly instead of return the `error`.

Reply:
Hey @rosstimothy it's a bit confusing but I think the original is the cleanest way to handle this, we can just rename the test error
```
	testErr = client.TestDraftExternalCloudAuditBuckets(ctx)
	diag, diagErr := s.handleErrFromBucketsTest(ctx, connectionDiagnosticID, testErr)
	if testErr != nil || diagErr != nil {
		return diag, diagErr
	}
```

`testErr` is not really an error for the sake of this function, it's just the subtest result, so I don't really consider it passing of error handling to another function. `handleErrFromBucketsTest` tries to append a diagnostic trace about whether the subtest passed or failed, but may itself fail to add the trace. We want to return early if either the subtest failed, or we failed to add the trace.

We could even name it `testResult` and `handleResultFromBucketsTest`